pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java

```diff
@@ -144,20 +141,34 @@ private QueryContext(@Nullable String tableName, @Nullable QueryContext subquery
     _explain = explain;
   }
 
+  public DataSourceContext getDataSource() {
+    return _dataSource;
+  }
+
   /**
    * Returns the table name.
-   * NOTE: on the broker side, table name might be {@code null} when subquery is available.
    */
   public String getTableName() {
-    return _tableName;
+    if (_dataSource.getSubquery() != null) {
+      return _dataSource.getSubquery().getTableName();
+    }
+    if (_dataSource.getTableName() != null) {
+      return _dataSource.getTableName();
+    }
+    assert _dataSource.getJoin() != null;
+    return _dataSource.getJoin().getLeftTableName();
```

Comment:
Not sure I follow this. Why should this return left table name in case of JOIN query ?